File: openai-intro/openai-intro.js
import OpenAI from 'openai';
import 'dotenv/config';

// Initialize OpenAI client
// Create an API key at https://platform.openai.com/api-keys
const client = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
});

console.log("=== OpenAI Intro: Understanding the Basics ===\n");

// ============================================
// EXAMPLE 1: Basic Chat Completion
// ============================================
async function basicCompletion() {
    console.log("--- Example 1: Basic Chat Completion ---");

    const response = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [
            { role: 'user', content: 'What is node-llama-cpp?' }
        ],
    });

    console.log("AI: " + response.choices[0].message.content);
    console.log("\n");
}

// ============================================
// EXAMPLE 2: Using System Prompts
// ============================================
async function systemPromptExample() {
    console.log("--- Example 2: System Prompts (Behavioral Control) ---");

    const response = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [
            { role: 'system', content: 'You are a coding assistant that talks like a pirate.' },
            { role: 'user', content: 'Explain what async/await does in JavaScript.' }
        ],
    });

    console.log("AI: " + response.choices[0].message.content);
    console.log("\n");
}

// ============================================
// EXAMPLE 3: Temperature and Creativity
// ============================================
async function temperatureExample() {
    console.log("--- Example 3: Temperature Control ---");

    const prompt = "Write a one-sentence tagline for a coffee shop.";

    // Low temperature = more focused and deterministic
    const focusedResponse = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.2,
    });

    // High temperature = more creative and varied
    const creativeResponse = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [{ role: 'user', content: prompt }],
        temperature: 1.5,
    });

    console.log("Low temp (0.2): " + focusedResponse.choices[0].message.content);
    console.log("High temp (1.5): " + creativeResponse.choices[0].message.content);
    console.log("\n");
}

// ============================================
// EXAMPLE 4: Conversation with Context
// ============================================
async function conversationContext() {
    console.log("--- Example 4: Multi-turn Conversation ---");

    // Build conversation history
    const messages = [
        { role: 'system', content: 'You are a helpful coding tutor.' },
        { role: 'user', content: 'What is a Promise in JavaScript?' },
    ];

    // First response
    const response1 = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: messages,
        max_tokens: 150,
    });

    console.log("User: What is a Promise in JavaScript?");
    console.log("AI: " + response1.choices[0].message.content);

    // Add AI response to history
    messages.push(response1.choices[0].message);

    // Add follow-up question
    messages.push({ role: 'user', content: 'Can you show me a simple example?' });

    // Second response (with context)
    const response2 = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: messages,
    });

    console.log("\nUser: Can you show me a simple example?");
    console.log("AI: " + response2.choices[0].message.content);
    console.log("\n");
}

// ============================================
// EXAMPLE 5: Streaming Responses
// ============================================
async function streamingExample() {
    console.log("--- Example 5: Streaming Response ---");
    console.log("AI: ");

    const stream = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [
            { role: 'user', content: 'Write a haiku about programming.' }
        ],
        stream: true,
    });

    for await (const chunk of stream) {
        const content = chunk.choices[0]?.delta?.content || '';
        process.stdout.write(content);
    }

    console.log("\n\n");
}

// ============================================
// EXAMPLE 6: Token Usage and Limits
// ============================================
async function tokenUsageExample() {
    console.log("--- Example 6: Understanding Token Usage ---");

    const response = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [
            { role: 'user', content: 'Explain recursion in 3 sentences.' }
        ],
        max_tokens: 100,
    });

    console.log("AI: " + response.choices[0].message.content);
    console.log("\nToken usage:");
    console.log("- Prompt tokens: " + response.usage.prompt_tokens);
    console.log("- Completion tokens: " + response.usage.completion_tokens);
    console.log("- Total tokens: " + response.usage.total_tokens);
    console.log("\n");
}

// ============================================
// EXAMPLE 7: Model Comparison
// ============================================
async function modelComparison() {
    console.log("--- Example 7: Different Models ---");

    const prompt = "What's 25 * 47?";

    // GPT-4o - Most capable
    const gpt4Response = await client.chat.completions.create({
        model: 'gpt-4o',
        messages: [{ role: 'user', content: prompt }],
    });

    // GPT-3.5-turbo - Faster and cheaper
    const gpt35Response = await client.chat.completions.create({
        model: 'gpt-3.5-turbo',
        messages: [{ role: 'user', content: prompt }],
    });

    console.log("GPT-4o: " + gpt4Response.choices[0].message.content);
    console.log("GPT-3.5-turbo: " + gpt35Response.choices[0].message.content);
    console.log("\n");
}

// ============================================
// Run all examples
// ============================================
async function main() {
    try {
        await basicCompletion();
        await systemPromptExample();
        await temperatureExample();
        await conversationContext();
        await streamingExample();
        await tokenUsageExample();
        await modelComparison();

        console.log("=== All examples completed! ===");
    } catch (error) {
        console.error("Error:", error.message);
        if (error.message.includes('API key')) {
            console.error("\nMake sure to set your OPENAI_API_KEY in a .env file");
        }
    }
}

main();

File: openai-intro/openai-intro.py
from openai import OpenAI
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Initialize OpenAI client
# Create an API key at https://platform.openai.com/api-keys
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

print("=== OpenAI Intro: Understanding the Basics ===\n")

# ============================================
# EXAMPLE 1: Basic Chat Completion
# ============================================
def basic_completion():
    print("--- Example 1: Basic Chat Completion ---")

    response = client.chat.completions.create(
        model='gpt-4o',
        messages=[
            {'role': 'user', 'content': 'What is llama-cpp-python?'}
        ],
    )

    print(f"AI: {response.choices[0].message.content}")
    print("\n")


# ============================================
# EXAMPLE 2: Using System Prompts
# ============================================
def system_prompt_example():
    print("--- Example 2: System Prompts (Behavioral Control) ---")

    response = client.chat.completions.create(
        model='gpt-4o',
        messages=[
            {'role': 'system', 'content': 'You are a coding assistant that talks like a pirate.'},
            {'role': 'user', 'content': 'Explain what async/await does in Python.'}
        ],
    )

    print(f"AI: {response.choices[0].message.content}")
    print("\n")


# ============================================
# EXAMPLE 3: Temperature and Creativity
# ============================================
def temperature_example():
    print("--- Example 3: Temperature Control ---")

    prompt = "Write a one-sentence tagline for a coffee shop."

    # Low temperature = more focused and deterministic
    focused_response = client.chat.completions.create(
        model='gpt-4o',
        messages=[{'role': 'user', 'content': prompt}],
        temperature=0.2,
    )

    # High temperature = more creative and varied
    creative_response = client.chat.completions.create(
        model='gpt-4o',
        messages=[{'role': 'user', 'content': prompt}],
        temperature=1.5,
    )

    print(f"Low temp (0.2): {focused_response.choices[0].message.content}")
    print(f"High temp (1.5): {creative_response.choices[0].message.content}")
    print("\n")


# ============================================
# EXAMPLE 4: Conversation with Context
# ============================================
def conversation_context():
    print("--- Example 4: Multi-turn Conversation ---")

    # Build conversation history
    messages = [
        {'role': 'system', 'content': 'You are a helpful coding tutor.'},
        {'role': 'user', 'content': 'What is a coroutine in Python?'},
    ]

    # First response
    response1 = client.chat.completions.create(
        model='gpt-4o',
        messages=messages,
        max_tokens=150,
    )

    print("User: What is a coroutine in Python?")
    print(f"AI: {response1.choices[0].message.content}")

    # Add AI response to history
    messages.append(response1.choices[0].message)

    # Add follow-up question
    messages.append({'role': 'user', 'content': 'Can you show me a simple example?'})

    # Second response (with context)
    response2 = client.chat.completions.create(
        model='gpt-4o',
        messages=messages,
    )

    print("\nUser: Can you show me a simple example?")
    print(f"AI: {response2.choices[0].message.content}")
    print("\n")


# ============================================
# EXAMPLE 5: Streaming Responses
# ============================================
def streaming_example():
    print("--- Example 5: Streaming Response ---")
    print("AI: ", end="")

    stream = client.chat.completions.create(
        model='gpt-4o',
        messages=[
            {'role': 'user', 'content': 'Write a haiku about programming.'}
        ],
        stream=True,
    )

    for chunk in stream:
        if chunk.choices[0].delta.content:
            print(chunk.choices[0].delta.content, end="", flush=True)

    print("\n\n")


# ============================================
# EXAMPLE 6: Token Usage and Limits
# ============================================
def token_usage_example():
    print("--- Example 6: Understanding Token Usage ---")

    response = client.chat.completions.create(
        model='gpt-4o',
        messages=[
            {'role': 'user', 'content': 'Explain recursion in 3 sentences.'}
        ],
        max_tokens=100,
    )

    print(f"AI: {response.choices[0].message.content}")
    print("\nToken usage:")
    print(f"- Prompt tokens: {response.usage.prompt_tokens}")
    print(f"- Completion tokens: {response.usage.completion_tokens}")
    print(f"- Total tokens: {response.usage.total_tokens}")
    print("\n")


# ============================================
# EXAMPLE 7: Model Comparison
# ============================================
def model_comparison():
    print("--- Example 7: Different Models ---")

    prompt = "What's 25 * 47?"

    # GPT-4o - Most capable
    gpt4_response = client.chat.completions.create(
        model='gpt-4o',
        messages=[{'role': 'user', 'content': prompt}],
    )

    # GPT-3.5-turbo - Faster and cheaper
    gpt35_response = client.chat.completions.create(
        model='gpt-3.5-turbo',
        messages=[{'role': 'user', 'content': prompt}],
    )

    print(f"GPT-4o: {gpt4_response.choices[0].message.content}")
    print(f"GPT-3.5-turbo: {gpt35_response.choices[0].message.content}")
    print("\n")


# ============================================
# Run all examples
# ============================================
def main():
    try:
        basic_completion()
        system_prompt_example()
        temperature_example()
        conversation_context()
        streaming_example()
        token_usage_example()
        model_comparison()

        print("=== All examples completed! ===")
    except Exception as error:
        print(f"Error: {str(error)}")
        if 'API key' in str(error):
            print("\nMake sure to set your OPENAI_API_KEY in a .env file")


if __name__ == "__main__":
    main()
